ggml-metal.metal

```diff
@@ -882,1012 +695,547 @@ typedef struct {
-        r[1] = (q[j+4] >>  4) | ((q[j-0] >> 6) << 4);
-        r[3] = (q[j+5] >>  4) | ((q[j+1] >> 6) << 4);
+//============================================ quant drivers ======================================================
+// load quantized blocks from device/threadgroup memory, dequantize 16 weights to a half4x4 or float4x4 type
+// init(il) : prepare some values that can be reused as long as il doesn't change.
+// dequantize(...) : dequantize 16 continuous weights.
+// inner_product_pre(il, yl) : multiply yl elements by a factor to speed up inner product calculations.
+// inner_product(...) : do inner product, may not use continuous weights.
+
+static inline void fix_y_v1(thread float & sumy, thread float4x4 & yl) {
+    sumy = 0.f;
+    for (int i = 0; i < 8; i += 2) {
+        sumy += yl[  i/4][i%4]; sumy += yl[  i/4][i%4+1];
+        sumy += yl[2+i/4][i%4]; sumy += yl[2+i/4][i%4+1];
+        yl[i/4  ][i%4  ] =            yl[  i/4][i%4];
+        yl[i/4  ][i%4+1] = 1/256.f  * yl[  i/4][i%4+1];
+        yl[i/4+2][i%4  ] = 1/16.f   * yl[2+i/4][i%4];
+        yl[i/4+2][i%4+1] = 1/4096.f * yl[2+i/4][i%4+1];
     }
-    return r;
 }
 
-//====================================== dot products =========================
-
-kernel void kernel_mul_mat_q2_K_f32(
-        device const  void * src0,
-        device const float * src1,
-        device       float * dst,
-        constant   int64_t & ne00,
-        constant   int64_t & ne01[[buffer(4)]],
-        constant   int64_t & ne02[[buffer(5)]],
-        constant   int64_t & ne10[[buffer(9)]],
-        constant   int64_t & ne12[[buffer(11)]],
-        constant   int64_t & ne0[[buffer(15)]],
-        constant   int64_t & ne1[[buffer(16)]],
-        constant   uint    & gqa[[buffer(17)]],
-        uint3 tgpig[[threadgroup_position_in_grid]],
-        uint tiisg[[thread_index_in_simdgroup]],
-        uint sgitg[[simdgroup_index_in_threadgroup]]) {
-
-    const int nb = ne00/QK_K;
-    const int r0 = tgpig.x;
-    const int r1 = tgpig.y;
-    const int r2 = tgpig.z;
-
-    const int first_row = (r0 * N_SIMDGROUP + sgitg) * N_DST;
-    const int ib_row = first_row * nb;
-    const uint offset0 = r2/gqa*(nb*ne0);
-    device const block_q2_K * x = (device const block_q2_K *) src0 + ib_row + offset0;
-    device const float      * y = (device const float      *) src1 + r1*ne10 + r2*ne00*ne1;
-    float yl[32];
-    float sumf[N_DST]={0.f}, all_sum;
+static inline void fix_y_v2(thread float & coef1, thread float & coef2, thread float & sumy, thread float4x4 & yl) {
+    sumy = 0.f;
+    for (int i = 0; i < 16; i += 2) {
+        sumy += yl[i/4][i%4];
+        sumy += yl[i/4][i%4+1];
+        yl[i/4][i%4]   = coef1 * yl[i/4][i%4];
+        yl[i/4][i%4+1] = coef2 * yl[i/4][i%4+1];
+    }
+}
 
-    const int step = sizeof(block_q2_K) * nb;
+template <typename addr_uint16_p,typename addr_block_q_p, typename type4x4>
+class q4_0_driver {
+    public:
+        uint16_t mask1, mask2, q_offset;
```

Comment:
I don't think we need high-level constructs for this type of computation code.

Having a `struct` with some data is fine, but all the functionality that operates on this data has to be imlpemented as plain functions with descriptive names.
Adding a `class` with methods here, introduces another level of indirection - when I see a call to `inner_product()`, I have to go back and look at the type of class. Instead, when I see for example `ggml_vec_dot_q4_0_q8_0()` I immediately know what is going on.